tests/components/tcp/test_sensor.py

```diff
@@ -36,228 +37,232 @@
-        """Set up things to be run when tests are started."""
-        self.hass = get_test_home_assistant()
 
-    def teardown_method(self, method):
-        """Stop everything that was started."""
-        self.hass.stop()
+@pytest.fixture
+def mock_select():
+    """Pytest fixture for socket."""
+    with patch("select.select", return_value=(True, False, False)) as mock_select:
+        yield mock_select
 
-    @patch("homeassistant.components.tcp.sensor.TcpSensor.update")
-    def test_setup_platform_valid_config(self, mock_update):
-        """Check a valid configuration and call add_entities with sensor."""
-        with assert_setup_component(0, "sensor"):
-            assert setup_component(self.hass, "sensor", TEST_CONFIG)
-
-        add_entities = Mock()
-        tcp.setup_platform(None, TEST_CONFIG["sensor"], add_entities)
-        assert add_entities.called
-        assert isinstance(add_entities.call_args[0][0][0], tcp.TcpSensor)
-
-    def test_setup_platform_invalid_config(self):
-        """Check an invalid configuration."""
-        with assert_setup_component(0):
-            assert setup_component(
-                self.hass, "sensor", {"sensor": {"platform": "tcp", "porrt": 1234}}
-            )
-
-    @patch("homeassistant.components.tcp.sensor.TcpSensor.update")
-    def test_name(self, mock_update):
-        """Return the name if set in the configuration."""
-        sensor = tcp.TcpSensor(self.hass, TEST_CONFIG["sensor"])
-        assert sensor.name == TEST_CONFIG["sensor"][tcp.CONF_NAME]
-
-    @patch("homeassistant.components.tcp.sensor.TcpSensor.update")
-    def test_name_not_set(self, mock_update):
-        """Return the superclass name property if not set in configuration."""
-        config = copy(TEST_CONFIG["sensor"])
-        del config[tcp.CONF_NAME]
-        entity = Entity()
-        sensor = tcp.TcpSensor(self.hass, config)
-        assert sensor.name == entity.name
-
-    @patch("homeassistant.components.tcp.sensor.TcpSensor.update")
-    def test_state(self, mock_update):
-        """Return the contents of _state."""
-        sensor = tcp.TcpSensor(self.hass, TEST_CONFIG["sensor"])
-        uuid = str(uuid4())
-        sensor._state = uuid
-        assert sensor.state == uuid
-
-    @patch("homeassistant.components.tcp.sensor.TcpSensor.update")
-    def test_unit_of_measurement(self, mock_update):
-        """Return the configured unit of measurement."""
-        sensor = tcp.TcpSensor(self.hass, TEST_CONFIG["sensor"])
-        assert (
-            sensor.unit_of_measurement
-            == TEST_CONFIG["sensor"][tcp.CONF_UNIT_OF_MEASUREMENT]
+
+async def test_setup_platform_valid_config(hass):
+    """Check a valid configuration and call add_entities with sensor."""
+    with assert_setup_component(0, "sensor"):
+        assert await async_setup_component(hass, "sensor", TEST_CONFIG)
+
+    add_entities = Mock()
+    tcp.setup_platform(None, TEST_CONFIG["sensor"], add_entities)
```

Comment:
If we set up the component the platform will be setup too. Either set up the platform directly, or better, just set up the component which will set up the platform and add an entity.

Reply:
I'm not 100% sure what your expectations are on this. I tried not to change the test logic from how it was before apart from transforming it the pytest method style, however, of course, I can provide improvments to this once I know how to handle this.

From what I understand from your comment and the code, `async_setup_component` is running `setup_platform` - correct? I think the reasoning that `setup_platform` is called again at this point is to verify it's call parameters. I'm not sure how to replicate this with using component setup. What is your expectation regarding this? Shall I drop the parameter validation and implement the test similar to e.g. `test_setup_platform_invalid_config`?

Reply:
We shouldn't do both approaches. That's wrong. Simplest fix is to remove the platform setup code and just get the state from the state machine, `hass.states`, and assert that there's a state for the corresponding entity id.

Reply:
Ok, thanks for clarifying. I'll look into changing the test that way.

Reply:
Reimplemented the test to check hass.states. Hope this is better now. However, I had to fall back to using the default template as for some reason, configuration checks wouldn't allow a `Template` object for `value_template`